File: workbook/ch08/data/disjoint.c
#include <stdio.h>

#define MAX 1000  // max number elements

typedef struct {
    int parent[MAX];
    int rank[MAX];
} DisjointSet;

void initialize(DisjointSet *ds, int n) {
    for (int i = 0; i < n; i++) {
        ds->parent[i] = i;  // each node is its own parent initially
        ds->rank[i] = 1;  // rank starts at 1
    }
}

int find(DisjointSet *ds, int x) {
    if (ds->parent[x] != x)
        ds->parent[x] = find(ds, ds->parent[x]);  // path compression
    return ds->parent[x];
}

void unionSets(DisjointSet *ds, int x, int y) {
    int rootX = find(ds, x);
    int rootY = find(ds, y);

    if (rootX != rootY) {
        if (ds->rank[rootX] > ds->rank[rootY]) {
            ds->parent[rootY] = rootX;
        } else if (ds->rank[rootX] < ds->rank[rootY]) {
            ds->parent[rootX] = rootY;
        } else {
            ds->parent[rootY] = rootX;
            ds->rank[rootX]++;
        }
    }
}

int connected(DisjointSet *ds, int x, int y) {
    return find(ds, x) == find(ds, y);
}

int main() {
    DisjointSet ds;
    int n = 10;
    initialize(&ds, n);

    unionSets(&ds, 1, 2);
    unionSets(&ds, 2, 3);
    unionSets(&ds, 4, 5);
    unionSets(&ds, 6, 7);

    printf("1 and 3 connected? %s\n", connected(&ds, 1, 3) ? "Yes" : "No");  // Yes
    printf("1 and 4 connected? %s\n", connected(&ds, 1, 4) ? "Yes" : "No");  // No

    unionSets(&ds, 3, 4);  // Connecting 3 and 4

    printf("1 and 4 connected after union(3,4)? %s\n", connected(&ds, 1, 4) ? "Yes" : "No");  // Yes

    return 0;
}

File: workbook/ch07/data/common/disjoint.py
class DisjointSet:
    def __init__(self, n):
        self.parent = [i for i in range(n)]  # Each node is its own parent initially
        self.rank = [1] * n  # Rank (size) for union by rank

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # Path compression
        return self.parent[x]

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)

        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.parent[rootY] = rootX
            elif self.rank[rootX] < self.rank[rootY]:
                self.parent[rootX] = rootY
            else:
                self.parent[rootY] = rootX
                self.rank[rootX] += 1  # Increase rank if they were of equal height

    def connected(self, x, y):
        return self.find(x) == self.find(y)

# example
ds = DisjointSet(10)
ds.union(1, 2)
ds.union(2, 3)
ds.union(4, 5)
ds.union(6, 7)

print("1 and 3 connected?", ds.connected(1, 3))  # True
print("1 and 4 connected?", ds.connected(1, 4))  # False

ds.union(3, 4)  # Connecting 3 and 4
print("1 and 4 connected after union(3,4)?", ds.connected(1, 4))  # True